include/deal.II/base/tensor.h

```diff
@@ -106,62 +348,80 @@ class Tensor
   typedef typename numbers::NumberTraits<Number>::real_type real_type;
 
   /**
+   * Type of objects encapsulated by this container and returned by
+   * operator[](). This is a tensor of lower rank for a general tensor, and
+   * a scalar number type for Tensor<1,dim,Number>.
+   */
+  typedef typename Tensor<rank_-1,dim,Number>::tensor_type value_type;
+
+  /**
    * Declare an array type which can be used to initialize an object of this
    * type statically.
    */
-  typedef typename Tensor<rank_-1,dim,Number>::array_type array_type[dim];
+  typedef typename Tensor<rank_-1,dim,Number>::array_type
+  array_type[(dim != 0) ? dim : 1];
+  // ... avoid a compiler warning in case of dim == 0 and ensure that the
+  // array always has positive size.
 
   /**
-   * Constructor. Initialize all entries to zero.
+   * Constructor. Initialize all entries to zero if
+   * <tt>initialize==true</tt>; this is the default behaviour.
    */
-  Tensor ();
+  explicit
+  Tensor (const bool initialize = true);
```

Comment:
If have removed the second recursion because it is unnecessary.

Maybe it is worthwhile to think about a more efficient allocator usage/ memset strategy in a separate pull request.